Feat/extensions sso (#2063)
mia-platform/documentation: docs/console/console-extensibility/sso.md

---
id: extension-sso
title: Extensions Single Sign On
sidebar_label: Extensions Single Sign On
---

# **Extensions Single Sign-On (SSO) Integration**

> The **Mia Platform Console** can serve as an **OAuth2 Authorization Server** for iframe-based extensions. When properly configured, this feature ensures that users logged into the Console are automatically authenticated in the extension eliminating the need for a second login.

Reviewer: ```suggestion
The **Mia Platform Console** can serve as an **OAuth2 Authorization Server** for iframe-based extensions. When properly configured, this feature ensures that users logged into the Console are automatically authenticated in the embedded extension, eliminating the need for a second login.
```

Following is described how this Single Sign-On (SSO) process works and how to integrate it into your iframe extension.  

### **Process Overview**

#### 1. **Initiate the Login Flow**

>    - The login flow begins when the user open the extension initial page. From this page of the extension the user is redirected to the following console URL that trigger the login process:

Reviewer: ```suggestion
   - The login flow begins when the user open the extension initial page. From this page of the extension the user is redirected to the following Console URL that triggers the login process:
```

     ```
     CONSOLE_BASE_URL/tenants/:tenantId/extension/:extensionId/authz?state={state}
     ```
     where
     - `state`: A value chosen by the developer to maintain session in the requests flow.
     - `tenantId`: ID of the Company
     - `extensionId` ID of the extension 

#### 2. **Redirect to the Extension Frontend**
   - As login process proceeds, the Mia Platform Console will redirect the user to the extension’s **callback URL**, appending the `code` and `state` query parameters.

#### 3. **Send Code and State from Frontend to Backend**

>    - The extension frontend has a page registerd on the *callback URL* route that receives the `code` and `state` query parameterrs. With these parameters the extension call its **backend** passing them for further processing.

Reviewer 2: ```suggestion
   - The extension frontend has a page registered on the *callback URL* route that receives the `code` and `state` query parameters. With these parameters, the extension calls its **backend** passing them for further processing.
```

#### 4. **Exchange Authorization Code for Access Token**

>    - The extension's backend then makes a POST request to the Mia Platform Console’s token endpoint utilizing `code` and `state` reccived:

Reviewer: ```suggestion
   - The extension's backend then makes a POST request to the Mia Platform Console’s token endpoint using `code` and `state` received:
```

     ```
     POST CONSOLE_BASE_URL/api/oauth/token
     ```
     - The request body should include the following:
       ```json
       {
         "code": "CODE",
         "state": "STATE"
       }
       ```

#### 5. **Receive Access Token**

>    - The console will respond with an **access token** that the extension backend return to its frontend and that can be uses to access protected resources in the backend.

Reviewer: ```suggestion
   - The Console will respond with an **access token** that the extension backend return to its frontend and that can be uses to access protected resources in the backend.
```

#### 6. **Validate the Access Token**
   - The extension backend can validate the token by calling the following Console endpoint:
     ```

>      GET CONSOLE_BASE_URL/jwks

Reviewer 2: there isn't any API prefix here?

     GET CONSOLE_BASE_URL/jwks
     ```

---

### **Setup Steps**

To complete the SSO integration, follow these setup steps:

#### 1. **Register an App for the Extension**

>    - Register the extension app via an **API call**

Author: here we add a link to the API documentation?

#### **Develop the Extension Frontend**

>   - Implement the redirect from the extension initial page to the console page that trigger the login process (see [Initiate the Login Flow](#1-initiate-the-login-flow)).

Reviewer: ```suggestion
  - Implement the redirect from the extension initial page to the Console page that trigger the login process (see [Initiate the Login Flow](#1-initiate-the-login-flow)).
```

  - Create a **callback page** within the extension’s frontend. This page should match the `callbackUrl` you defined during app registration. It will handle the `code` and `state` query parameters received from the redirect.

#### **Develop the Extension Backend**

>   - Implement an endpoint in the extension's backend (e.g., `/token`) that communicates with the console’s `CONSOLE_BASE_URL/api/oauth/token` endpoint. This allows the backend to exchange the authorization code for an access token.

Reviewer: ```suggestion
  - Implement an endpoint in the extension's backend (e.g., `/token`) that communicates with the Console’s `CONSOLE_BASE_URL/api/oauth/token` endpoint. This allows the backend to exchange the authorization code for an access token.
```

---

> By following these steps, you can seamlessly implement **Single Sign-On (SSO)** for your extensions using Mia Platform’s OAuth2 Authorization Server, ensuring secure user authentication and access to protected backend resources.

Reviewer: could be useful to insert a link to the Oauth2 framework documentation?

The Mia-Platform Console can serve as an **OAuth2 Authorization Server** for iframe-based extensions. When properly configured, this feature ensures that users logged into the Console are automatically authenticated in the embedded extension, eliminating the need for a second login.

Here below you can read how to enable this feature via Console and how the flow for the SSO integration works within your iframe extension.

## Feature Overview

With SSO enabled, users who are authenticated in the Mia-Platform Console do not need to log in again when accessing the extension embedded within the Console. This seamless experience improves user convenience and security by leveraging OAuth2 authentication.

To enable and configure this feature, you can use the Console interface, specifically within the extension’s detail section. After enabling the feature, you will provide a **Callback URL**, which is the route in your extension that will receive the OAuth2 authorization code for further processing and where the Console will redirect the user after completing the login process.

The steps to enable SSO on your extension are therefore the following:

> - Go to the detail page of your iframe-based extension.

Author: here would be useful a image of the new section
